Profile the fetch phase

nik9000 opened the pull request:
This adds profiling to the fetch phase so we can tell when fetching is
slower than we'd like and we can tell which portion of the fetch is
slow. The output includes which stored fields were loaded, how long it
took to load stored fields, which fetch sub-phases were run, and how
long those fetch sub-phases took.

Closes #75892

nik9000 committed:
Profile the fetch phase

This adds profiling to the fetch phase so we can tell when fetching is
slower than we'd like and we can tell which portion of the fetch is
slow. The output includes which stored fields were loaded, how long it
took to load stored fields, which fetch sub-phases were run, and how
long those fetch sub-phases took.

Closes #75892

nik9000 commented on client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/search/TransportNoopSearchAction.java:
```diff
@@ -40,7 +40,7 @@ protected void doExecute(Task task, SearchRequest request, ActionListener<Search
             InternalAggregations.EMPTY,
             new Suggest(Collections.emptyList()),
-            new SearchProfileShardResults(Collections.emptyMap()), false, false, 1),
+            new SearchProfileResults(Collections.emptyMap()), false, false, 1),
```
Sorry for the renames. The old names were confusing and I introduces a couple of new classes which would have been even more confusing without the rename.

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -163,7 +163,38 @@ The API returns the following result:
+        "aggregations": [],
+        "fetch": {
+          "type": "fetch",
+          "description": "fetch",
```
This `type` and `description` come from my using a raw `ProfileResult` for the entire fetch phase. I'm not happy about them, but I wasn't sure if it was worth writing a whole top level class for the result or working around them some other way.

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -163,7 +163,38 @@ The API returns the following result:
+        "fetch": {
+          "type": "fetch",
+          "description": "fetch",
+          "time_in_nanos": 660555,
```
This is the actual `nanoTime` for the entire fetch phase rather than the sum of the breakdown. Because we can do that in fetch. We couldn't really do that for aggs or queries.

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -163,7 +163,38 @@ The API returns the following result:
+            "load_stored_fields_count": 5
+          },
+          "debug": {
+            "stored_fields": ["_id", "_routing", "_source"]
```
This bit right here is important to me. Some folks want to only fetch doc values and avoid stored fields entirely. With this PR you can profile the query and check that this array is empty. If it is then we skipped loading doc values. Unless something tricky is happening in the nested case that I didn't dig into.

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -163,7 +163,38 @@ The API returns the following result:
+          "children": [
+            {
+              "type": "source",
+              "description": "load _source",
```
We get one of these per fetch phase we run.

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -163,7 +163,38 @@ The API returns the following result:
+              },
+              "debug": {
+                "loaded_nested": 0,
+                "fast_path": 5
```
The fetch phases can emit extra debugging information if they decide to do so. This has proven super useful in aggs land. I'm not sure how useful it is here, but I figured it'd be nice to be able to glance at a profile result and know if we were able to just shovel the bytes out the other side or if we had to parse the source.

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -673,13 +704,15 @@ The API returns the following result:
 }
 --------------------------------------------------
 // TESTRESPONSE[s/"aggregations": \[\.\.\.\]/"aggregations": $body.$_path/]
+// TESTRESPONSE[s/"fetch": \{\.\.\.\}/"fetch": $body.$_path/]
```
If we like this PR I'll update the docs to add a section describing the fetch profile. The guts of a profile are in a kind of grey area where they are quite part of the public API but folks can look at them and kibana even visualizes then but we will change them when the underlying implementations change. Its a weird place.

nik9000 commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorHighlightSubFetchPhase.java:
```diff
@@ -40,6 +40,16 @@
+    @Override
+    public String name() {
+        return "percolator_highlight";
+    }
```
These power the `type` and `description` fields on the `children` field in the fetch profile.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -78,11 +79,27 @@ public void execute(SearchContext context) {
             return;
         }
 
+        Profiler profiler = context.getProfilers() == null ? Profiler.NOOP : context.getProfilers().getFetchProfiler();
```
I wasn't super happy with how invasive the change was here. I'm happy that the sub-phases for the most part don't have to change which is good, but this bit is bigger than I'd like. I'd love any suggestions if anyone has any.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -428,4 +475,57 @@ private static void fillDocAndMetaFields(SearchContext context, FieldsVisitor fi
         return docs.length > 0 && docs[docs.length-1].docId - docs[0].docId == docs.length - 1;
     }
+
+    interface Profiler {
```
I went with the internal interface and noop implementation to save from a bunch of null checks that'd have made the whole thing even more invasive. I'd be more than happy to get rid of them if anyone has any good ideas.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourcePhase.java:
```diff
@@ -50,46 +62,52 @@ public void process(HitContext hitContext) {
-    @SuppressWarnings("unchecked")
-    private void hitExecute(FetchSourceContext fetchSourceContext, HitContext hitContext) {
+            @SuppressWarnings("unchecked")
+            private void hitExecute(FetchSourceContext fetchSourceContext, HitContext hitContext) {
```
Pulled into the phase object so we could collect debugging counters.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/ScriptFieldsPhase.java:
```diff
@@ -22,10 +22,19 @@
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext context) {
-        if (context.scriptFields() == null) {
+        if (context.scriptFields() == null || context.scriptFields().fields().isEmpty()) {
```
When you don't ask for any script fields we tend to get an empty list here, not null. That empty list was causing us to always return the phase and run it. That's not a big deal at all because it just iterates an empty list, but it was clogging up the profile output.

nik9000 reviewed (commented):
I implemented this by having the fetch phase optionally profile itself, attaching the profile results to the per-shard fetch results. I then merge those fetch results into the "search phase" profiling that we have been collecting for years. The rendered json of the profile result will only include a `fetch` section on shards that performed a fetch.

nik9000 committed:
Skip bwc

nik9000 commented:
>  ** elasticsearch-ci/part-1 ** — Build finished.

I can't reproduce this failure. Its in the CCS dueling tests. Both look like they fetch the same 8 documents but they fetch them from different shards: https://gist.github.com/nik9000/809fa7865e3d5b1c7dd50e96b89e3452

nik9000 commented:
> I can't reproduce this failure. Its in the CCS dueling tests. Both look like they fetch the same 8 documents but they fetch them from different shards: https://gist.github.com/nik9000/809fa7865e3d5b1c7dd50e96b89e3452

Nope. I was on the wrong branch. I can reproduce. Checking.

nik9000 commented:
> Nope. I was on the wrong branch. I can reproduce. Checking.

OK! It's because the test assumed that the entire search result would be the same when searching in minimize round trip mode vs non-minimized mode. If you minimize the round trips you will perform more fetches so, of course, there will be more fetches.

nik9000 committed:
Don't compare fetch profiles

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Use passed one

nik9000 committed:
no npe

nik9000 commented:
So folks don't have to dig through the docs, this is what the fetch profile looks like:
```
        "fetch": {
          "type": "fetch",
          "description": "fetch",
          "time_in_nanos": 660555,
          "breakdown": {
            "next_reader": 7292,
            "next_reader_count": 1,
            "load_stored_fields": 299325,
            "load_stored_fields_count": 5
          },
          "debug": {
            "stored_fields": ["_id", "_routing", "_source"]
          },
          "children": [
            {
              "type": "source",
              "description": "load _source",
              "time_in_nanos": 20443,
              "breakdown": {
                "next_reader": 745,
                "next_reader_count": 1,
                "process": 19698,
                "process_count": 5
              },
              "debug": {
                "loaded_nested": 0,
                "fast_path": 5
              }
            }
          ]
        }
```

nik9000 committed:
Do last rename

nik9000 committed:
Move method down

nik9000 commented:
I've found a few more unit tests I'd like to write. Incoming.

nik9000 committed:
serialization tests

nik9000 committed:
Fix sneaky serialization

nik9000 committed:
Test for sneaky bug

nik9000 committed:
license header

nik9000 commented:
> I've found a few more unit tests I'd like to write. Incoming.

500 lines of tests..... The profile stuff didn't have any round trip unit tests so I added a bunch. Those are useful to catch serialization issues super early.

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Document

nik9000 committed:
Fix test

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
newline

jtibshirani commented on qa/multi-cluster-search/src/test/java/org/elasticsearch/search/CCSDuelIT.java:
```diff
@@ -813,6 +817,8 @@ private static void assertAggs(SearchResponse response) {
             List<Map<String, Object>> shards = (List <Map<String, Object>>)profile.get("shards");
             for (Map<String, Object> shard : shards) {
                 replaceProfileTime(shard);
+                // Fetch profiling will be significantly different in fan out mode
```
Could you explain this part? I'm not totally clear on why it would be different.

nik9000 replied:
I'll add an explanation, yeah. The short version is that what minimizing round trips does is that it'll fetch on the remote cluster and pull back the response and then merge *those* responses. In full fan out mode we'll merge the candidates before the fetching.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -148,22 +172,24 @@ public void execute(SearchContext context) {
                 throw new FetchPhaseExecutionException(context.shardTarget(), "Error running fetch phase for doc [" + docId + "]", e);
             }
         }
+        for (FetchSubPhaseProcessor processor : processors) {
```
Do we need this new method `done`? It looks like the only implementation adds the debug info, which seems like it could be done when initializing the processor:

```
    @Override
            public void done() {
                breakdown.debug = delegate.getDebugInfo();
            }
```

nik9000 replied:
I'll double check. I sure thought I needed it but I may not.

nik9000 replied:
You were right. I didn't need it.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -241,9 +268,9 @@ private HitContext prepareHitContext(SearchContext context,
+                context, profiler, fieldsVisitor, docId, storedToRequestedFields, subReaderContext, storedFieldReader);
         } else {
-            return prepareNestedHitContext(context, docId, nestedDocuments, storedToRequestedFields,
+            return prepareNestedHitContext(context, profiler, docId, nestedDocuments, storedToRequestedFields,
```
It would be nice to have a test or example of how the output looks when there are nested hits.

nik9000 replied:
++

nik9000 replied:
I've added support for nested and parent/child based inner_hits - you see the whole sub-sub-fetch phase as a child of the phase. It looks like:
```
  1>             "fetch" : {
  1>               "type" : "fetch",
  1>               "description" : "fetch",
  1>               "time_in_nanos" : 4181937,
  1>               "breakdown" : {
  1>                 "load_stored_fields" : 133903,
  1>                 "load_stored_fields_count" : 1,
  1>                 "next_reader" : 13716,
  1>                 "next_reader_count" : 1
  1>               },
  1>               "debug" : {
  1>                 "stored_fields" : [
  1>                   "_id",
  1>                   "_routing",
  1>                   "_source"
  1>                 ]
  1>               },
  1>               "children" : [
  1>                 {
  1>                   "type" : "inner_hits",
  1>                   "description" : "fetches matching nested or parent or child documents",
  1>                   "time_in_nanos" : 3046811,
  1>                   "breakdown" : {
  1>                     "process_count" : 1,
  1>                     "process" : 3046193,
  1>                     "next_reader" : 618,
  1>                     "next_reader_count" : 1
  1>                   },
  1>                   "children" : [
  1>                     {
  1>                       "type" : "question",
  1>                       "description" : "inner_hits",
  1>                       "time_in_nanos" : 101738,
  1>                       "breakdown" : {
  1>                         "load_stored_fields" : 23394,
  1>                         "load_stored_fields_count" : 1,
  1>                         "next_reader" : 2734,
  1>                         "next_reader_count" : 1
  1>                       },
  1>                       "debug" : {
  1>                         "stored_fields" : [
  1>                           "_id",
  1>                           "_routing",
  1>                           "_source"
  1>                         ]
  1>                       },
  1>                       "children" : [
  1>                         {
  1>                           "type" : "source",
  1>                           "description" : "load _source",
  1>                           "time_in_nanos" : 4243,
  1>                           "breakdown" : {
  1>                             "process_count" : 1,
  1>                             "process" : 3949,
  1>                             "next_reader" : 294,
  1>                             "next_reader_count" : 1
  1>                           },
  1>                           "debug" : {
  1>                             "fast_path" : 1,
  1>                             "loaded_nested" : 0
  1>                           }
  1>                         }
  1>                       ]
  1>                     }
  1>                   ]
  1>                 },

```

It just exposes whatever is in the sub-phase without thinking about what the sub-phase tries to share. Hopefully that should all come out in the timings.

nik9000 replied:
I've reverted this bit so we can do it in a follow up.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -390,11 +431,17 @@ private HitContext prepareNestedHitContext(SearchContext context,
-        fieldVisitor.reset();
-        fieldReader.accept(docId, fieldVisitor);
-        fieldVisitor.postProcess(fieldTypeLookup);
+        try {
```
As a heads up, we've brainstormed moving stored fields loading to its own subphase. I'm not sure it will happen, but this would affect the output structure (which is fine, it's not actually part of our API).

nik9000 replied:
++ - would we want to do only a single stored field load?

jtibshirani replied:
Yep I think that'd be the goal.

jtibshirani reviewed (commented):
This is looking like a great change! I left some initial comments as I got up-to-speed. I had some higher-level thoughts:
* Part of the "invasiveness" of the change comes from the top-level fetch profiling, for example profiling the combined `setNextReader` time. I assume this is valuable vs. just adding the individual timings together as we do for the query phase? 
* The change is large so it's hard to be fully confident I'm not missing something. Unfortunately we don't have great test coverage for the fetch phase, especially source loading and inner hits, making it important to review carefully. Is it possible to break up the PR to make it easier to review? Or maybe you could highlight any parts that could possibly affect the core fetch logic?

nik9000 commented:
>     * Part of the "invasiveness" of the change comes from the top-level fetch profiling, for example profiling the combined `setNextReader` time. I assume this is valuable vs. just adding the individual timings together as we do for the query phase?

I added the top level `setNextReader` because we do "stuff" in there in the top level fetch phase. I expect it to always be fast, but I figure the whole point of this change is to catch unexpected things.


>     * The change is large so it's hard to be fully confident I'm not missing something. Unfortunately we don't have great test coverage for the fetch phase, especially source loading and inner hits, making it important to review carefully. Is it possible to break up the PR to make it easier to review? Or maybe you could highlight any parts that could possibly affect the core fetch logic?

I'm sure I can slice out the renames without much trouble. I could slice out the new test cases I added for serialization for existing classes too.

I could slice out bits of the fetch profiling itself, though none of it feels *that* big. Do you have ideas on how to do it? Or should I try slicing out the renames and the tests to see where that lands us?

jtibshirani commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search/330_fetch_fields.yml:
```diff
@@ -892,3 +892,57 @@ Test token_count inside nested field doesn't fail:
+
+
+---
+profile:
```
I don't feel strongly but have a slight preference for adding a new yml test focused on profiling.

nik9000 replied:
I've moved to a `profile` yaml file.

jtibshirani commented:
> I could slice out bits of the fetch profiling itself, though none of it feels that big. Do you have ideas on how to do it? Or should I try slicing out the renames and the tests to see where that lands us?

I think just pulling out the renames and additional tests would help a lot. That would make it easier to stare deeply at the fetch-related changes.

nik9000 commented:
> I think just pulling out the renames and additional tests would help a lot. That would make it easier to stare deeply at the fetch-related changes.

I've opened #77195.

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Restore assertion

nik9000 commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/integration/DocumentLevelSecurityTests.java:
```diff
@@ -1400,13 +1402,15 @@ public void testProfile() throws Exception {
+        assertThat(
+            queryProfileShardResult.getQueryResults().stream().map(ProfileResult::getLuceneDescription).sorted().collect(toList()),
+            equalTo(List.of("(other_field:value)^0.8"))
+        );
```
This had been commented out for years. I'm not sure why. It passes quite consistently for me. I think its worth uncommenting it and watching.

nik9000 replied:
Hopefully folks'll git-blame it and come for me. But yeah, I'll ping them before merging.

nik9000 reviewed (commented):
Now that I've removed the renames I hope this is a little clearer. Now this change:
1. Creates `SearchProfileQueryPhaseResults` which contains *just* `SearchProfileQueryPhaseResult` collected during the query phase. After the fetch phase we `merge` the fetch results into it to produce the full `SearchProfileResult`.
2. Creates `SearchProfileShardResult` which contains a `SearchProfileQueryPhaseResult` and an optional `ProfileResult` for the fetch. This now lives in the `SearchProfileResult` and is produced by the `merge` called after the fetch phase is complete.

nik9000 committed:
unit test merging

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Handle inner hits

nik9000 commented:
@jtibshirani I believe this should be ready for another round.

If you want, there are two more ways we could shrink it:
1. Move the new tests for old objects into their own PR.
2. Move the two places where I move a method into the sub-phase-processor into their own PRs.

nik9000 committed:
Fixup

nik9000 commented:
run elasticsearch-ci/part-2 

We've lost the CI run and I have no idea what failed.

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Revert unneeded

jtibshirani commented on server/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java:
```diff
@@ -40,16 +43,24 @@ public InternalSearchResponse(SearchHits hits, InternalAggregations aggregations
         );
     }
 
+    private static SearchProfileResults readProfile(StreamInput in) throws IOException {
```
Do we need this special handling if we already have BWC checks in `SearchProfileResults(StreamInput)`?

nik9000 replied:
I'm going to have to drop some breakpoints on this one. It sure doesn't look like we should need it. It'd be nice to be able to remove `Writeable` from `SearchProfileQueryPhaseResults`. I thought we did, but I'll be we do not.

nik9000 replied:
I've figured it out - you were right - I was doing the bwc logic twice. Actually using one way on read and one on write. Bad idea! I'm dropping this method entirely and letting the `SearchProfileResults` handle the read.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileQueryPhaseResults.java:
```diff
@@ -0,0 +1,123 @@
+/**
+ * Profile results for the query phase run on all shards.
+ */
+public final class SearchProfileQueryPhaseResults implements Writeable { // This is Writeable for backwards compatibility
```
For me it'd be clearer to call this `SearchProfileResultsBuilder` and rename `merge` -> `build`. It took me a while to wrap my head around all the classes.

nik9000 replied:
I hadn't thought of that name! I'll give it more thought because I'm kind of used to the old name. I understand wanting to change it, yeah.

jtibshirani commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileQueryPhaseResults.java:
```diff
@@ -0,0 +1,123 @@
+     * @return A {@link SearchProfileQueryPhaseResult} representing the results for this
+     *         shard
+     */
+    public static SearchProfileQueryPhaseResult buildShardResults(Profilers profilers) {
```
Maybe this should live on `SearchProfileQueryPhaseResult` (singular).

nik9000 replied:
I was thinking of maybe moving it into it's call site. I don't think it really fit into its old spot either.

nik9000 replied:
I think I'll move it into `Profilers`.

jtibshirani reviewed (commented):
I did a full review round and left a few comments!

One overall comment: I had several questions around the API and implementation for inner_hits, and think that it'd be worth pulling that implementing out into its own PR. I have some ideas for how to factor it to make it less invasive, but (as we discussed earlier) this code is subtle and doesn't have great test coverage. This PR is already quite large and I think it'd be great to tackle inner_hits in a follow-up.

nik9000 commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -575,8 +592,18 @@ static int getTopDocsSize(SearchRequest request) {
-        public InternalSearchResponse buildResponse(SearchHits hits) {
-            return new InternalSearchResponse(hits, aggregations, suggest, shardResults, timedOut, terminatedEarly, numReducePhases);
+        public InternalSearchResponse buildResponse(SearchHits hits, Collection<? extends SearchPhaseResult> fetchResults) {
+            SearchProfileResults profileResults = mergeProfile(fetchResults);
```
I'm sort of worried that adding too much merging logic to this class'd make it harder to think about because it's super big and has different threads coming at it. I really just pulled this out to make the null guard more obvious. I'd be ok doing whatever you think is right here.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/FetchSubPhase.java:
```diff
@@ -74,6 +74,16 @@ public IndexReader topLevelReader() {
+    /**
+     * Name of the phase to include in profile results.
+     */
+    String name();
```
I could do the `getSimpleName` thing. I was thinking of doing that at first but it felt a little too "cute". But if two of us thought about it its probably ok. I'll change to that and remove the new methods.

nik9000 commented on server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourcePhase.java:
```diff
@@ -50,46 +62,52 @@ public void process(HitContext hitContext) {
+                // Otherwise, filter the source and add it to the hit.
+                Object value = source.filter(fetchSourceContext);
+                if (nestedHit) {
+                    loadedNested++;
```
The more I look at the code the more I think we don't really need it. You can tell from where you are in the process if you loaded a nested object. But maybe this isn't as clear to someone who's not super into the fetch phase stuff? I dunno. I think I'd be happy to drop it.

nik9000 commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileShardResult.java:
```diff
@@ -0,0 +1,98 @@
+ * Profile results from a particular shard for all search phases.
+ */
+public class SearchProfileShardResult implements Writeable, ToXContentFragment {
+    private final SearchProfileQueryPhaseResult search;
```
:+1: that name is a leftover. I renamed the object but not the field or the method to read it.

nik9000 commented on server/src/test/java/org/elasticsearch/action/search/SearchPhaseControllerTests.java:
```diff
@@ -221,13 +230,26 @@ public void testMerge() {
-                0, true, InternalAggregationTestCase.emptyReduceContextBuilder(), true);
-            AtomicArray<SearchPhaseResult> fetchResults = generateFetchResults(nShards,
-                reducedQueryPhase.sortedTopDocs.scoreDocs, reducedQueryPhase.suggest);
+            AtomicArray<SearchPhaseResult> queryResults = generateQueryResults(nShards, suggestions, queryResultSize, false, profile);
```
Yes. If it isn't in the loop and we profile then the test fails because we consume the profile twice.

nik9000 commented on server/src/test/java/org/elasticsearch/search/fetch/FetchProfilerTests.java:
```diff
@@ -0,0 +1,53 @@
+        assertMap(copy.getDebugInfo(), matchesMap().entry("stored_fields", List.of("_id", "_routing", "_source")));
+    }
+
+    static class FixedTimeProfiler extends FetchProfiler {
```
:+1:

nik9000 reviewed (commented):
I'd be happy to pull the inner hits stuff into a follow up. I could disable profiling on the "inner fetch phase" that it runs - you'd still get profiling on the inner hits sub-phase.

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Revert inner hits profiling

nik9000 committed:
Fix names

nik9000 committed:
Fixup names

nik9000 committed:
Move results building

nik9000 commented:
I'm pushing some updates that handle most of your easy comments. I believe this is what is left:
* [ ] Figure out if `mergeProfile` is good
* [ ] Figure out the BWC wire protocol thing
* [ ] Decide if I like renaming to `SearchProfileResultsBuilder`

nik9000 committed:
Drop loaded_nested

nik9000 committed:
Checkstyle

nik9000 committed:
Merge branch 'master' into profile_fetch_simple

nik9000 committed:
Fixup more

nik9000 committed:
Finish writeable cleanup

Add unit tests for merge

nik9000 committed:
Remove null checking builder

nik9000 commented:
@jtibshirani  I think this is ready for another round! Presuming tests pass but things are looking fairly good locally.

I renamed the holder for query phase results into `SearchProfileResultsBuilder` and moved all of the logic I could into it.

I ended up removing `mergeProfile` because I figured out a way I could unit test it in isolation. It's basically just a ternary but I'm happy any time I can unit test a small chunk of `SearchPhaseController`.

I removed the double-bwc handling. That was a silly mistake.

nik9000 commented:
> @jtibshirani I think this is ready for another round! Presuming tests pass but things are looking fairly good locally.

And there's a failure! Looking

nik9000 committed:
Fix wire mistake

How did this pass before?!

nik9000 commented:
> And there's a failure! Looking

It was a serialization error. The right unit test caught it. Odd that it didn't cause havoc, but it was simple enough to fix.

jtibshirani commented on server/src/test/java/org/elasticsearch/search/profile/SearchProfileQueryPhaseResultsTests.java:
```diff
@@ -0,0 +1,101 @@
+import static org.hamcrest.Matchers.matchesPattern;
+import static org.hamcrest.Matchers.nullValue;
+
+public class SearchProfileQueryPhaseResultsTests extends ESTestCase {
```
I think this should be named `SearchProfileResultsBuilderTests` now.

nik9000 replied:
:+1:

jtibshirani commented on server/src/main/java/org/elasticsearch/search/profile/SearchProfileResultsBuilder.java:
```diff
@@ -0,0 +1,88 @@
+ * Profile results for the query phase run on all shards.
+ */
+public class SearchProfileResultsBuilder {
+    public static SearchProfileResultsBuilder build(Map<String, SearchProfileQueryPhaseResult> queryPhaseResults) {
```
I am a little confused about this part -- probably related to naming. Could we just have a constructor `new SearchProfileResultsBuilder(Map<String, SearchProfileQueryPhaseResult>)` instead of this static method? I'm not sure why we need the special `NOT_PROFILING` object instead of just passing `null` in the original place in `SearchPhaseController`.

nik9000 replied:
I was trying to get the null check out of search phase controller so it was easy to unit test without all of SearchPhaseController. This way there isn't a null check. I liked that the old code didn't allocate anything if we didn't profile and wanted to keep that too, but have it here.

Another idea is to make the `build` method static and have it take a nullable builder param. Or just revert the whole change and keep the null check in SearchPhaseController. I just, like, didn't like it there. I'm happy anytime I can get things *out* of SearchPhaseController.

jtibshirani replied:
Since this change has so many "moving parts" already, I'd be in favor of keeping the old strategy to keep it well-scoped. I also find it easier to understand. If we want we could follow-up with a refactor to simplify `SearchPhaseController` a bit.

nik9000 replied:
Yeah. I think it was clearer that way but I'm not the only one that has to read it. I've changed it the way you asked.

jtibshirani reviewed (commented):
It is so close! I had one last comment.

nik9000 committed:
Rename

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/FetchSubPhaseProcessor.java:
```diff
@@ -28,4 +31,18 @@
+    /**
+     * Profiles from child-phases.
+     */
+    default List<ProfileResult> childProfiles() {
```
No current subprocessor overrides this, could we remove until a follow-up where we implement profiling for inner_hits subphases?

jtibshirani commented on server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java:
```diff
@@ -153,17 +176,17 @@ public void execute(SearchContext context) {
                 FetchSubPhaseProcessor processor = fsp.getProcessor(context);
                 if (processor != null) {
-                    processors.add(processor);
+                    String type = fsp.getClass().getSimpleName().replaceAll("^Fetch", "").replaceAll("(FetchSub)?Phase$", "");
```
Personally I think it's nice to simply include the class name without replacements. It's still very readable and makes it easier to find the corresponding code for the phase.

jtibshirani reviewed (commented):
I was wrong about that being the last comment. I have a few more small ones.

nik9000 commented:
Sure! That's cool. I'll go back to the null check method, remove the name
munging, and blast the leftover for child phases. I just missed it last
time.

On Fri, Sep 10, 2021, 13:30 Julie Tibshirani ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> I was wrong about that being the last comment. I have a few more small
> ones.
> ------------------------------
>
> In
> server/src/main/java/org/elasticsearch/search/fetch/FetchSubPhaseProcessor.java
> <https://github.com/elastic/elasticsearch/pull/77064#discussion_r706344964>
> :
>
> > @@ -28,4 +31,18 @@
>       */
>      void process(HitContext hitContext) throws IOException;
>
> +    /**
> +     * Called when profiling after processing all documents to get any extra
> +     * debug information the phase collected.
> +     */
> +    default Map<String, Object> getDebugInfo() {
> +        return null;
> +    }
> +
> +    /**
> +     * Profiles from child-phases.
> +     */
> +    default List<ProfileResult> childProfiles() {
>
> No current subprocessor overrides this, could we remove until a follow-up
> where we implement profiling for inner_hits subphases?
> ------------------------------
>
> In server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java
> <https://github.com/elastic/elasticsearch/pull/77064#discussion_r706346307>
> :
>
> >          try {
>              List<FetchSubPhaseProcessor> processors = new ArrayList<>();
>              for (FetchSubPhase fsp : fetchSubPhases) {
>                  FetchSubPhaseProcessor processor = fsp.getProcessor(context);
>                  if (processor != null) {
> -                    processors.add(processor);
> +                    String type = fsp.getClass().getSimpleName().replaceAll("^Fetch", "").replaceAll("(FetchSub)?Phase$", "");
>
> Personally I think it's nice to simply include the class name without
> replacements. It's still very readable and makes it easier to find the
> corresponding code for the phase.
> ------------------------------
>
> In
> server/src/main/java/org/elasticsearch/search/profile/SearchProfileResultsBuilder.java
> <https://github.com/elastic/elasticsearch/pull/77064#discussion_r706353912>
> :
>
> > +
> +package org.elasticsearch.search.profile;
> +
> +import org.elasticsearch.search.SearchPhaseResult;
> +import org.elasticsearch.search.fetch.FetchSearchResult;
> +
> +import java.util.Collection;
> +import java.util.Collections;
> +import java.util.HashMap;
> +import java.util.Map;
> +
> +/**
> + * Profile results for the query phase run on all shards.
> + */
> +public class SearchProfileResultsBuilder {
> +    public static SearchProfileResultsBuilder build(Map<String, SearchProfileQueryPhaseResult> queryPhaseResults) {
>
> Since this change has so many "moving parts" already, I'd be in favor of
> keeping the old strategy to keep it well-scoped. I also find it easier to
> understand. If we want we could follow-up with a refactor to simplify
> SearchPhaseController a bit.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/77064#pullrequestreview-751629710>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABUXIU5N3HB3ODBAS2RE7TUBI6D5ANCNFSM5DEDX3OQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

nik9000 committed:
Remove funny builder

nik9000 committed:
Remove name munging

nik9000 commented:
OK! Back to you @jtibshirani.

jtibshirani reviewed (approved):
Looks good to me!

nik9000 commented:
In and backported! Exciting!

javanna commented:
Heya, shall we update the limitations from the profile docs page now that fetch can be profiled too? https://www.elastic.co/guide/en/elasticsearch/reference/current/search-profile.html#profile-limitations

nik9000 commented:
> Heya, shall we update the limitations from the profile docs page now that fetch can be profiled too? https://www.elastic.co/guide/en/elasticsearch/reference/current/search-profile.html#profile-limitations

I have no idea when, but someone's removed that "limitation".
